### DISCUSSION TITLE
Weighted calculations

### DISCUSSION THREAD

**Discussion by danielcoral** [MSKCC-Epi-Bio/decisioncurveanalysis#5] [Ideas]
Hi all,

Thanks for designing such a useful tool to assess clinical relevance of biomarkers.
I was wondering if a weighted calculation of DCA could be derived, so that each individual contributes to the calculation by a certain weight. This means that one should be able to specify weights when calculating rates. For example, the `.surv_to_risk` function used to calculate disease rates in the time-to-event setting could look like this:

```
# mat would be here a model matrix that includes a "weights" column
.surv_to_risk <- function(mat, time){
    survival:::survfit.formula(
        mat[,1] ~ 1, 
        data = mat,
        weights = mat[,"weights"]
    ) %>%
        broom::tidy() %>%
        dplyr::filter(time <= time) %>%
        dplyr::mutate(estimate = 1 - estimate) %>%
        dplyr::slice_tail() %>%
        dplyr::pull(estimate)
}
```

Thanks again!

Daniel.

**Comment by VickersA** [comment]
Daniel: weights cannot be directly specified in the current decision curve analysis code and we have no plans to add this as a feature. Part of the reason is that there is a trivial work around. Because there is no concern about the standard error for most implementations of decision curve analysis, you can simply create an arbitrarily large data set by weighted sampling.

1.	Format your data set so that it has three columns: probability from model (from 0 to 1), true outcome (either 0 or 1), weight (from 0 to 1). Note that, obviously, if it is survival time data, it will be four columns with indicator and time instead of 0/1 true outcome.
2.	Copy the data set to a new file. 
3.	Assign a random number from 0 to 1 to each observation. 
4.	Delete the observation if the random number is greater than the weight. 
5.	Save the data set, which will have a smaller sample size than the original. 
6.	Repeat steps 2 – 5 a large number of times (e.g. at least 500, but it shouldn’t really matter with contemporary hardware, even 10,000 iterations should take only a few minutes). 
7.	Combine all the data sets created in steps 2 – 5 into one data set. 
8.	Run the decision curve analysis on the combined data set. 

Hope this helps. 

Andrew
